execute/tokendata/observer_background.go

```diff
@@ -0,0 +1,431 @@
+func (q *msgQueue) containsMsg(msg cciptypes.Message) bool {
+	q.mu.RLock()
+	defer q.mu.RUnlock()
+
+	for _, qMsg := range q.msgs {
+		if qMsg.msg.Header.MessageID == msg.Header.MessageID {
+			q.lggr.Debugw("message already exists in the queue", "msg", msg, "qMsg", qMsg)
+			return true
+		}
+	}
+
+	return false
+}
+
+// size returns the number of messages in the queue
+func (q *msgQueue) size() int {
+	q.mu.RLock()
+	defer q.mu.RUnlock()
+	return len(q.msgs)
+}
+
+type inMemTokenDataCache struct {
+	lggr               logger.Logger
+	expirationInterval time.Duration
+	inMemTokenData     map[cciptypes.Bytes32]exectypes.MessageTokenData
+	expiresAt          map[cciptypes.Bytes32]time.Time
+	mu                 *sync.RWMutex
+	done               chan struct{}
+}
+
+// newInMemObservationsCache initializes an in-memory cache for token data.
+// It uses a background goroutine to periodically check and remove expired data.
+// cleanupInterval specifies the frequency for checking and cleaning up inactive data.
+// Setting a low value is discouraged, as the cleanup process holds a lock.
+func newInMemObservationsCache(lggr logger.Logger,
+	expirationInterval, cleanupInterval time.Duration, doneChan chan struct{}) *inMemTokenDataCache {
+	c := &inMemTokenDataCache{
+		lggr:               lggr,
+		expirationInterval: expirationInterval,
+		inMemTokenData:     make(map[cciptypes.Bytes32]exectypes.MessageTokenData),
+		expiresAt:          make(map[cciptypes.Bytes32]time.Time),
+		mu:                 &sync.RWMutex{},
+		done:               doneChan,
+	}
+	c.runExpirationLoop(cleanupInterval)
+	return c
+}
+
+// get returns the token data for the given message ID if it exists in the cache.
+func (c *inMemTokenDataCache) get(msgID cciptypes.Bytes32) (exectypes.MessageTokenData, bool) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	msgData, ok := c.inMemTokenData[msgID]
+	if !ok {
+		return exectypes.MessageTokenData{}, false
+	}
+
+	return msgData, true
+}
+
+// set stores the token data in memory with the given expiration interval.
+func (c *inMemTokenDataCache) set(msgID cciptypes.Bytes32, tokenData exectypes.MessageTokenData) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.inMemTokenData[msgID] = tokenData
+	c.expiresAt[msgID] = time.Now().Add(c.expirationInterval).UTC()
+	c.lggr.Debugw("token data cached", "msgID", msgID, "expiresAt", c.expiresAt[msgID])
+}
+
+// size returns the number of cached token data
+func (c *inMemTokenDataCache) size() int {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return len(c.inMemTokenData)
+}
+
+// runExpirationLoop is a background goroutine that periodically checks and removes expired data.
+func (c *inMemTokenDataCache) runExpirationLoop(cleanupInterval time.Duration) {
+	go func() {
+		ticker := time.NewTicker(cleanupInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-c.done:
+				c.lggr.Debug("expiration loop gracefully stopped")
+				return
+			case <-ticker.C:
```

Comment:
Why not use `cache.Cache` for that? It's widely adopted in our codebase and probably has all the required features to handle that, no?